src/lib/spinel/spinel_driver.cpp

```diff
@@ -0,0 +1,473 @@
+#include "lib/platform/exit_code.h"
+#include "lib/spinel/spinel.h"
+
+namespace ot {
+namespace Spinel {
+
+constexpr spinel_tid_t sTid = 1; ///< In Spinel Driver, only use Tid as 1.
+
+SpinelDriver::SpinelDriver(void)
+    : Logger("SpinelDriver")
+    , mSpinelInterface(nullptr)
+    , mWaitingKey(SPINEL_PROP_LAST_STATUS)
+    , mIsWaitingForResponse(false)
+    , mIid(SPINEL_HEADER_INVALID_IID)
+    , mSpinelVersionMajor(-1)
+    , mSpinelVersionMinor(-1)
+    , mIsCoProcessorReady(false)
+{
+    memset(mVersion, 0, sizeof(mVersion));
+
+    mReceivedFrameHandler = &HandleInitialFrame;
+    mFrameHandlerContext  = this;
+}
+
+void SpinelDriver::Init(SpinelInterface    &aSpinelInterface,
+                        bool                aSoftwareReset,
+                        const spinel_iid_t *aIidList,
+                        uint8_t             aIidListLength)
+{
+    mSpinelInterface = &aSpinelInterface;
+    mRxFrameBuffer.Clear();
+    SuccessOrDie(mSpinelInterface->Init(HandleReceivedFrame, this, mRxFrameBuffer));
+
+    VerifyOrDie(aIidList != nullptr, OT_EXIT_INVALID_ARGUMENTS);
+    VerifyOrDie(aIidListLength != 0 && aIidListLength <= mIidList.GetMaxSize(), OT_EXIT_INVALID_ARGUMENTS);
+
+    for (uint8_t i = 0; i < aIidListLength; i++)
+    {
+        SuccessOrDie(mIidList.PushBack(aIidList[i]));
+    }
+    mIid = aIidList[0];
+
+    ResetCoProcessor(aSoftwareReset);
+    SuccessOrDie(CheckSpinelVersion());
+    SuccessOrDie(GetCoprocessorVersion());
+    SuccessOrDie(GetCoprocessorCaps());
+}
+
+void SpinelDriver::Deinit(void)
+{
+    // This allows implementing pseudo reset.
+    new (this) SpinelDriver();
+}
+
+otError SpinelDriver::SendReset(uint8_t aResetType)
+{
+    otError        error = OT_ERROR_NONE;
+    uint8_t        buffer[kMaxSpinelFrame];
+    spinel_ssize_t packed;
+
+    // Pack the header, command and key
+    packed = spinel_datatype_pack(buffer, sizeof(buffer), SPINEL_DATATYPE_COMMAND_S SPINEL_DATATYPE_UINT8_S,
+                                  SPINEL_HEADER_FLAG | SPINEL_HEADER_IID(mIid), SPINEL_CMD_RESET, aResetType);
+
+    VerifyOrExit(packed > 0 && static_cast<size_t>(packed) <= sizeof(buffer), error = OT_ERROR_NO_BUFS);
+
+    SuccessOrExit(error = mSpinelInterface->SendFrame(buffer, static_cast<uint16_t>(packed)));
+    LogSpinelFrame(buffer, static_cast<uint16_t>(packed), true /* aTx */);
+
+exit:
+    return error;
+}
+
+void SpinelDriver::ResetCoProcessor(bool aSoftwareReset)
+{
+    bool hardwareReset;
+    bool resetDone = false;
+
+    // Avoid resetting the device twice in a row in Multipan RCP architecture
+    VerifyOrExit(!mIsCoProcessorReady, resetDone = true);
+
+    mWaitingKey = SPINEL_PROP_LAST_STATUS;
+
+    if (aSoftwareReset && (SendReset(SPINEL_RESET_STACK) == OT_ERROR_NONE) && (!mIsCoProcessorReady) &&
+        (WaitResponse() == OT_ERROR_NONE))
+    {
+        VerifyOrExit(mIsCoProcessorReady, resetDone = false);
+        LogCrit("Software reset Co-Processor successfully");
```

Comment:
nit: several more instances below
```suggestion
        LogCrit("Software reset co-processor successfully");
```